lib/rosette/integrations/smartling-integration/translation_memory.rb

```diff
@@ -1,43 +1,237 @@
+          else
+            resolve_without_placeholders(variant)
+          end
+        end
+
+        def source_placeholders_for(phrase)
+          phrase.key.scan(placeholder_regex)
+        end
+
+        def target_placeholders_for(variant)
+          variant.elements.each_with_object([]) do |el, ret|
+            case el
+              when TmxParser::Placeholder
+                ret << el.text
+              else
+                str = el.respond_to?(:text) ? el.text : el
+                ret.concat(find_inline_placeholders(str))
+            end
+          end
+        end
+
+        def associate_placeholders(target_placeholders, source_placeholders)
+          Hash[target_placeholders.zip(source_placeholders)]
+        end
+
+        def resolve_with_placeholders(variant, placeholders)
+          variant.elements.each_with_object('') do |el, ret|
+            ret << case el
+              when TmxParser::Placeholder
+                placeholders[el.text] || ''
+              else
+                str = el.respond_to?(:text) ? el.text : el
+
+                if str
+                  str.dup.tap do |text|
+                    placeholders.each do |source, target|
+                      text.sub!(source, target) if source && target
+                    end
+                  end
+                else
+                  ''
+                end
+            end
+          end
+        end
+
+        def resolve_without_placeholders(variant)
+          variant.elements.each_with_object('') do |el, ret|
+            ret << (el.respond_to?(:text) ? el.text : el)
+          end
+        end
+
+        def digest_locale(locale, digest)
+          translations = all_translations_for(locale)
+
+          translations.keys.sort.each do |meta_key|
+            digest_string(meta_key, digest)
+
+            translations[meta_key].each do |unit|
+              digest_unit(unit, locale, digest)
+            end
+          end
+        end
+
+        def digest_string(string, digest)
+          digest << string
+        end
+
+        def digest_unit(unit, locale, digest)
+          variants = unit.variants.sort { |v1, v2| v1.locale <=> v2.locale }
+          variants.each do |variant|
+            variant.elements.each do |element|
+              digest << case element
+                when String
+                  element
+                else
+                  element.text
+              end
+            end
+          end
+        end
+
+        def find_variant(unit, locale)
+          unit.variants.find do |variant|
+            variant.locale == locale.code || variant.locale == locale.language
+          end
+        end
+
+        def is_potential_plural?(meta_key)
+          !!(meta_key =~ PLURAL_REGEX)
+        end
+
+        def find_plural_translation_for(locale, meta_key)
+          all_translations = all_translations_for(locale)
+          meta_key_base = meta_key.sub(PLURAL_REGEX, '')
+          plural_form = $1
```

Comment:
because this is so far (code wise) from the definition of the regex, consider adding a comment here about what will be returned in the $1 capture